beetsplug/fish.py

```diff
@@ -0,0 +1,245 @@
+# This file is part of beets.
+# Copyright 2015, Franois-Xavier Thomas.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""If you use the fish-shell http://fishshell.com/ ... this will do
+autocomplete for you.
+It does the main commands and options for beet and the plugins.
+It gives you all the album and itemfields (like genre, album) but
+not all the values for these. It suggest genre: or album:
+but not genre: Pop..Jazz...Rock
+You can get that by specifying ex. --extrafields genre.
+"""
+
+from __future__ import (division, absolute_import, print_function,
+                        unicode_literals)
+
+from beets.plugins import BeetsPlugin
+from beets import library, ui
+from beets.ui import commands
+from operator import attrgetter
+import os
+
+
+class fishPlugin(BeetsPlugin):
+    def commands(self):
+        cmd = ui.Subcommand('fish', help='make fish autocomplete beet')
+        cmd.func = self.run
+        cmd.parser.add_option('-f', '--noFields', action='store_true',
+                              default=False,
+                              help='add fields')
+        cmd.parser.add_option('-e', '--extraFields',
+                              action='append',
+                              type='choice',
+                              choices=library.Item.all_keys() +
+                              library.Album.all_keys(),
+                              help='extra fields to autocomplete')
+        return [cmd]
+
+    def run(self, lib, opts, args):
+        homeDir = os.path.expanduser("~")
+        completePath = os.path.join(homeDir, '.config/fish/completions')
+        try:
+            os.makedirs(completePath)
+        except OSError:
+            if not os.path.isdir(completePath):
+                raise
+        pathAndFile = os.path.join(completePath, 'beet.fish')
+        with open(pathAndFile, 'w') as fish_file:
+            fish_file.write(get_fishfile(opts, lib))
```

Comment:
Rather than passing around `opts` to the various utility functions, it can make for more readable code if you split out the specific options into their own variables here. Then the `run` function is the only one that has to deal with wrangling user-supplied arguments, and the rest of the code can be "safer" because it's accessing local variables directly rather than looking up in an environment.